### PR TITLE
FEATURE: Collapsable sidebar

### PR DESCRIPTION
**Pull request by keegangeorge**
Adds support for default category logos when logos are not present. Also adds the ability to toggle to collapse/expand sidebar to show icons only or full widget.

### CODE REVIEW

File: common/common.scss [hunk 1]
```diff
@@ -94,6 +106,20 @@
     }
   }
 
+  ul.parent-categories .layouts-minimize-button {
+    margin-top: 2rem;
+    padding-top: 0.25rem;
+    padding-bottom: 0.25rem;
+    border-top: 2px solid var(--primary-low);
+    .minimize-icon {
```
Comment (angusmcleod) [inline common/common.scss:119]: Missing a line here (before child rule)

File: javascripts/discourse/widgets/layouts-category-list.js [hunk 2]
```diff
@@ -186,6 +190,54 @@ export default layouts.createLayoutsWidget('category-list', {
...
+  tagName: 'li.layouts-minimize-button.layouts-category-link',
+  buildKey: (attrs) => 'layouts-minimize-categories',
+  
+  defaultState(attrs) {
+    return {
+      sidebarMinimized: false
+    }
+  },
+
+  html(attrs, state) {
```
Comment (angusmcleod) [inline javascripts/discourse/widgets/layouts-category-list.js:204]: Considering we're always going to have the same result here, we should perhaps simplify this a bit, i.e.

```
html() {
   return [
      h('div.category-logo.minimize-icon', iconNode('chevron-circle-left')),
      h('div.category-name.minimize-text', I18n.t(themePrefix("minimize_button_label")))
   ]
}
```

We can always add in more logic later, however if you look at the code now you're first given the impression there must be some other logic there, but it turns out it's always the same two elements.

File: javascripts/discourse/widgets/layouts-category-list.js [hunk 3]
```diff
@@ -186,6 +190,54 @@ export default layouts.createLayoutsWidget('category-list', {
...
+  
+  click(attrs) {
+    const categoryDetails = document.querySelectorAll('.layouts-category-link .category-name, .layouts-category-link .d-icon-lock');
+    const sidebar = document.querySelector('aside.sidebar');
+    const minimizeIcon = document.querySelector('.minimize-icon');
+    const siteSettings = this.register.lookup("site-settings:main");
+    const sidebarLeftWidth = siteSettings.layouts_sidebar_left_width;
+    
+    if (!this.state.sidebarMinimized) {
+      categoryDetails.forEach((category) => {
```
Comment (angusmcleod) [inline javascripts/discourse/widgets/layouts-category-list.js]: @keegangeorge Like with jquery we should only be using javascript to change the classList outside of the render thread of the frameworks we're using if it's the last option available. Rather we should be using:

1. Events to send actions up the widget hierarchy
2. Attributes to send values and current state down the widget hiearchy

Then updating the classlists of various widgets or their html using ``attrs`` and ``state``. Only if that's not possible do we resort to updating the DOM directly like this. Same in Ember as well. Try to use the framework first and direct class manipulation like this as a last resort.

If you're not sure what I'm talking about here, let's chat about it in our catchup this week.

**Review by angusmcleod — CHANGES_REQUESTED** [review]

**Comment by keegangeorge** [comment]
Improved and moved some functionality directly to layouts plugin so it can be used by other widgets as well.
See PR: https://github.com/paviliondev/discourse-layouts/pull/16

**Pull request merged**